apps/ejabberd/src/mod_aws_sns.erl

```diff
@@ -0,0 +1,293 @@
+    ejabberd_hooks:add(user_available_hook, Host, ?MODULE, user_present, 90),
+    ejabberd_hooks:add(unset_presence_hook, Host, ?MODULE, user_not_present, 90),
+
+    application:ensure_all_started(erlcloud),
+    application:ensure_all_started(worker_pool),
+
+    WorkerNum = gen_mod:get_opt(pool_size, Opts, 100),
+    wpool:start_sup_pool(pool_name(Host), [{workers, WorkerNum}]),
+
+    %% Check for required options
+    RequiredOptions = [access_key_id, secret_access_key, region, account_id, sns_host],
+    lists:foreach(
+        fun(Option) ->
+            case opt(Host, Option) of
+                undefined ->
+                    error({missing_required_option, Option});
+                _ -> ok
+            end
+         end, RequiredOptions),
+
+    ok.
+
+-spec stop(Host :: ejabberd:server()) -> ok.
+stop(Host) ->
+    MUCHost = gen_mod:get_module_opt_subhost(Host, ?MODULE, mod_muc:default_host()),
+    ejabberd_hooks:delete(unset_presence_hook, Host, ?MODULE, user_not_present, 90),
+    ejabberd_hooks:delete(user_available_hook, Host, ?MODULE, user_present, 90),
+    ejabberd_hooks:delete(user_send_packet, Host, ?MODULE, user_send_packet, 90),
+    ejabberd_hooks:delete(rest_user_send_packet, Host, ?MODULE, user_send_packet, 90),
+    ejabberd_hooks:delete(filter_room_packet, MUCHost, ?MODULE, filter_room_packet, 90),
+
+    wpool:stop(pool_name(Host)),
+
+    ok.
+
+
+%% Handle user_send_packet hook and muc messages
+-spec user_send_packet(From :: ejabberd:jid(), To :: ejabberd:jid(),
+                       Packet :: jlib:xmlel()) -> 'ok'.
+user_send_packet(From = #jid{lserver = Host}, To, Packet) ->
+    ?DEBUG("SNS Packet handle~n    from ~p ~n    to ~p~n    packet ~p.", [From, To, Packet]),
+
+    case {get_topic(Host, Packet), exml_query:subelement(Packet, <<"body">>)} of
+        {undefined, _} -> %% Skip if there is no topic set in configuration for the packet type
+            skip;
+        {_, undefined} -> %% Skip if there is no message body in the packet
+            skip;
+        {Topic, BodyTag} ->
+            FromGUID = user_guid(Host, From),
+            ToGUID = user_guid(Host, To),
+            MessageBody = exml_query:cdata(BodyTag),
+            Content = #{from_user_id => FromGUID,
+                        to_user_id => ToGUID,
+                        message => MessageBody},
+
+            TopicARN = make_topic_arn(Host, Topic),
+            Attributes = message_attributes(Host, TopicARN, From, To, message_type(Packet), Packet),
+
+            async_publish(Host, TopicARN, Content, Attributes)
+    end,
+
+    ok.
+
+%% Handle filter_room_packet
+-spec filter_room_packet(Packet :: jlib:xmlel(), EventData :: proplists:proplist()) ->
+    Packet :: jlib:xmlel().
+filter_room_packet(Packet, EventData) ->
+    {_, FromJID} = lists:keyfind(from_jid, 1, EventData),
+    {_, RoomJID} = lists:keyfind(room_jid, 1, EventData),
+    catch user_send_packet(FromJID, RoomJID, Packet),
```

Comment:
It states what is does. It handles filter_room_packet hook. Every single module in MongooseIM uses hook name as function name even if this function has side effect which in most cases - it does.